cardano-db-sync/src/Cardano/DbSync/Cache/FIFO.hs

```diff
@@ -0,0 +1,144 @@
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+
+module Cardano.DbSync.Cache.FIFO (
+  FifoCache (..),
+  TxIdCache,
+  HashableTxId (..),
+  empty,
+  insert,
+  lookup,
+  getSize,
+  getCapacity,
+  emptyTxIdCache,
+  insertTxIdCache,
+  lookupTxIdCache,
+  cleanupTxIdCache,
+) where
+
+import Cardano.Crypto.Hash.Class (hashToBytes)
+import qualified Cardano.Db as DB
+import Cardano.Ledger.Crypto (StandardCrypto)
+import Cardano.Ledger.SafeHash (extractHash)
+import Cardano.Ledger.TxIn (TxId (..))
+import Cardano.Prelude hiding (empty)
+import qualified Data.HashMap.Strict as HM
+import Data.Hashable (Hashable, hashWithSalt)
+import qualified Data.List.NonEmpty as NE
+
+{- FifoCache: A high-performance, fixed-capacity FIFO (First-In-First-Out) cache implementation
+
+Performance characteristics:
+1. Space efficiency:
+   - Uses UNPACK pragma for maxCapacity and currentSize, reducing memory overhead
+   - Compact representation using NonEmpty for the queue and HashMap for the cache
+
+2. Time complexity:
+   - Lookup: O(1) average case, leveraging HashMap's constant-time lookups
+   - Insert: O(1) amortized time
+     - HashMap insert/delete: O(1) average case
+     - NonEmpty cons/uncons: O(1)
+   - When cache is full:
+     - Efficiently removes oldest item using NonEmpty's O(1) uncons operation
+     - Maintains FIFO order without full list traversal
+
+3. Optimizations:
+   - Uses strict fields (!) to prevent space leaks
+   - Leverages NonEmpty for the queue, ensuring type-safe non-empty invariant
+   - Employs HashMap for fast key-value lookups and updates
+
+4. Scalability:
+   - Performs well with increasing cache sizes due to HashMap's efficiency
+   - Constant-time operations regardless of cache occupancy
+
+5. Memory usage:
+   - Bounded by maxCapacity, preventing unbounded growth
+   - UNPACK pragma reduces memory footprint for frequently accessed fields
+
+6. Thread safety:
+   - Note: This implementation is not thread-safe. Concurrent access should be
+     managed externally if required.
+-}
+
+data FifoCache k v = FifoCache
+  { maxCapacity :: {-# UNPACK #-} !Natural
+  , currentSize :: {-# UNPACK #-} !Int
+  , queue :: !(Maybe (NE.NonEmpty k))
+  , cacheMap :: !(HM.HashMap k v)
```

Comment:
Any specific reason to use HashMap instead of simple Map? Won't this cause hash collisions?

Reply:
it's because for lookups, `HashMap` is O(1) average case, O(n) worst case yet `Map` is O(log n) which is useful for large dataset. 

`HashMap` will use slightly more memory usage but the performance gains arise when dealing upwards of tens of thousands of elements.

There shouldn't be any hash collisions as from my understanding all `txLedgerTxId` are unique by default.

Reply:
`txLedgerTxId` are unique, but the hash of the tx is rehashed by the HashMap implementation. This second hash is not collision resistant and is one of the reasons that HashMap's are never used in the codebase of Cardano.

Existance of collisions is documented at https://hackage.haskell.org/package/hashable-1.5.0.0/docs/Data-Hashable.html#g:1

Reply:
ahh wondered why there wasn't a Hashable instance of `TxId`.

Well that throws a spanner in the works. Guess I'll just remove `queue :: !(Maybe (NE.NonEmpty k))` and make the cache using `Data.Map.Strict`. I was trying to be as fast as possible 😓

Reply:
This shouldn't affect queue. Can't Hashable simply change to Data.Map.Strict?

Reply:
yeah that's what I've done now saves all the custom instances and the `Hashable` issue:

```
data FIFOCache k v = FIFOCache
  { maxCapacity :: !Word64
  , cacheMap :: !(Map.Map k v)
  , keyOrder :: !(Seq.Seq k)  -- To track insertion order
  }

empty :: Word64 -> FIFOCache k v
empty capacity = FIFOCache capacity Map.empty Seq.empty

insert :: Ord k => k -> v -> FIFOCache k v -> FIFOCache k v
insert key value cache@FIFOCache{..}
  | Map.size cacheMap >= fromIntegral maxCapacity =
      let oldestKey = Seq.index keyOrder 0
          !newKeyOrder = Seq.drop 1 keyOrder Seq.|> key
          !newMap = Map.insert key value $ Map.delete oldestKey cacheMap
      in cache { cacheMap = newMap, keyOrder = newKeyOrder }
  | otherwise =
      cache { cacheMap = Map.insert key value cacheMap
            , keyOrder = keyOrder Seq.|> key }

lookup :: Ord k => k -> FIFOCache k v -> Maybe v
lookup key FIFOCache{..} = Map.lookup key cacheMap

getSize :: FIFOCache k v -> Int
getSize = Map.size . cacheMap

getCapacity :: FIFOCache k v -> Word64
getCapacity = maxCapacity

cleanupCache :: FIFOCache k v -> FIFOCache k v
cleanupCache cache = cache { cacheMap = Map.empty, keyOrder = Seq.empty }
```

Reply:
Seems, good, except I would suggest using the Strict version of Sequence in `cardano-strict-containers` `Data.Sequence.Strict`

Reply:
cool just added `Data.Sequence.Strict` think I have it right!
```haskell
data FIFOCache k v = FIFOCache
  { maxCapacity :: !Word64
  , cacheMap :: !(Map.Map k v)
  , keyOrder :: !(StrictSeq k)
  }

empty :: Word64 -> FIFOCache k v
empty capacity = FIFOCache capacity Map.empty Seq.empty

insert :: Ord k => k -> v -> FIFOCache k v -> FIFOCache k v
insert key value cache@FIFOCache {..}
  | Map.size cacheMap >= fromIntegral maxCapacity =
      case Seq.lookup 0 keyOrder of
        Nothing -> cache -- This should never happen if invariants are maintained
        Just oldestKey ->
          let !newKeyOrder = key Seq.<| Seq.drop 1 keyOrder
              !newMap = Map.insert key value $ Map.delete oldestKey cacheMap
           in cache {cacheMap = newMap, keyOrder = newKeyOrder}
  | otherwise =
      cache
        { cacheMap = Map.insert key value cacheMap
        , keyOrder = keyOrder Seq.|> key
        }

lookup :: Ord k => k -> FIFOCache k v -> Maybe v
lookup key FIFOCache {..} = Map.lookup key cacheMap

getSize :: FIFOCache k v -> Int
getSize = Map.size . cacheMap

getCapacity :: FIFOCache k v -> Word64
getCapacity = maxCapacity

cleanupCache :: FIFOCache k v -> FIFOCache k v
cleanupCache cache = cache {cacheMap = Map.empty, keyOrder = Seq.empty}
```